src/plugins/discord/mirror.js

```diff
@@ -0,0 +1,111 @@
+// @flow
+
+import {type DiscordFetcher, type ResultPage} from "./fetcher";
+import {type SqliteMirror} from "./sqliteMirror";
+import {type Snowflake} from "./models";
+import * as Model from "./models";
+import * as nullUtil from "../../util/null";
+
+type FetchType<T> = (Snowflake) => Promise<ResultPage<T>>;
+
+/**
+ * Mirrors data from the Discord API into a local sqlite db.
+ */
+export class Mirror {
+  +_guildId: Snowflake;
+  +_sqliteMirror: SqliteMirror;
+  +_fetcher: DiscordFetcher;
+
+  constructor(
+    guildId: Snowflake,
+    sqliteMirror: SqliteMirror,
+    fetcher: DiscordFetcher
+  ) {
+    this._guildId = guildId;
+    this._sqliteMirror = sqliteMirror;
+    this._fetcher = fetcher;
+  }
+
+  async _paginatedFetch<T>(
+    fetch: (endCursor: Snowflake) => Promise<ResultPage<T>>,
+    process: (object: T) => void | Promise<void>,
+    endCursor: Snowflake
+  ): Promise<void> {
+    let after = endCursor;
+    let hasNextPage = true;
+
+    while (hasNextPage) {
+      const {results, pageInfo} = await fetch(after);
+      for (const result of results) {
+        await process(result);
+      }
+      hasNextPage = pageInfo.hasNextPage;
+      if (hasNextPage) {
+        after = nullUtil.get(
+          pageInfo.endCursor,
+          `Found null endCursor with hasNextPage = true`
+        );
+      }
+    }
+  }
+
+  async _fetchMembers(): Promise<void> {
+    const fetchMembers: FetchType<Model.GuildMember> = async (
+      after: Snowflake
+    ) => {
+      return await this._fetcher.members(this._guildId, after);
+    };
+
+    const processMember: (Model.GuildMember) => void = (
+      member: Model.GuildMember
+    ) => this._sqliteMirror.addMember(member);
+
+    await this._paginatedFetch(fetchMembers, processMember, "0");
```

Comment:
I tried passing `this._sqliteMirror.addMember` directly, instead of writing `processMember`; I was getting an error in the `sqliteMirror` about `this` being `none`, where `this` should have referenced the `sqliteMirror`.

Reply:
Yeah, if you call the function outside of it's object context, `this` will be unset. There's a few ways around that using any of `Function.prototype.apply`, `call` or `bind`. Bind being easiest but also performance heavy, so avoid using in hot loops.

```js
myObj.func();
// this == myObj

let func = myObj.func;
func();
// this == undefined

func = myObj.func.bind(myObj);
func();
// this == myObj
```